### PR TITLE
New Console Class && Logging + ClientDir changes

### PR DESCRIPTION
**Pull request by UnseenFaith**
### Proposed Semver Increment Bump: [MAJOR/MINOR/PATCH]
Patch
### Changes Proposed in this Pull Request (List new items in CHANGELOG.MD)

New Things:
 - New Console Class
 - Breaking changes in config (moved them around)
 - No longer need Chalk ~~(Once finalizer is fixed)~~
 - HexString and HexNumber Format color support (0x and #)
- RGB Color Format [number, number, number]
- HSL Color Format [string, string%, string%]

Subject to change:
 - ~~Color String Names (light instead of bright)~~ light instead of bright now
 - Snake_Case Strings? 
 - UPPERCASE instead of lowercase?
 - RGB/HSL Format... rgb(r,g,b) && hsl(h,s,l) as strings might be more efficient?

Things Left to do: 
- ~~Fix CommandLogging Finalizer (fucking forgot this uses chalk as well..)~~
- ~~Add 88/256 Color Support~~
- ~~Fix Internal Attribution code ( I complicated it a lot. )~~
- ~~Maybe HSL support?~~

Disclaimer: Some things will work and some things won't work. This is entirely based upon what terminal you use. If something isn't as the documentation says, then chances are the terminal you're using changed it. It will not be changed as the current implementation is the most static one across all terminals.
d4b0d1655ed4c6463907855c3f8a1c34

EDIT from **Cyberium**: Attached my ClientDir changes to this PR as my changes were checkedout from this PR

### CODE REVIEW

File: src/classes/console/Console.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+  }
+
+  /**
+   * Flattens our data into a readable string.
+   * @param  {*} data Some data to flatten
+   * @return {string}
+   */
+  static flatten(data) {
+    data = data.stack || data.message || data;
+    if (typeof data === "object" && typeof data !== "string" && !Array.isArray(data)) data = require("util").inspect(data, { depth: 0, colors: true });
```
Comment (kyranet) [inline src/classes/console/Console.js]: require("util") should be outside (no-inline-require) and the colors option should be editable.

**Review by CyberiumShadow — DISMISSED** [review]
Code-wise looks good to me. If tested feel free to go ahead

File: src/classes/console/Console.js [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+     * @name KomadaConsole#stderr
+     * @type {WritableStream}
+     */
+    Object.defineProperty(this, "stderr", { value: stderr });
+
+    /**
+     * Whether or not timestamps should be enabled for this console.
+     * @type {boolean}
+     */
+    this.timestamps = timestamps;
```
Comment (kyranet) [inline src/classes/console/Console.js]: (Port from Klasa):
```javascript
this.timestamps = timestamps === true ? "YYYY-MM-DD HH:mm:ss" : timestamps;
```

File: src/classes/console/Console.js [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+  /**
+   * Logs everything to the console/writable stream.
+   * @param  {*} stuff The stuff we want to print.
+   * @param  {string} [type="log"] The type of log, particularly useful for coloring.
+   */
+  log(stuff, type = "log") {
+    stuff = KomadaConsole.flatten(stuff, this.useColors);
+    const message = this.colors ? this.colors[type.toLowerCase()].message : {};
+    const time = this.colors ? this.colors[type.toLowerCase()].time : {};
+    const timestamp = this.timestamps ? this.timestamp(`[${moment().format("YYYY-MM-DD HH:mm:ss")}]`, time) : "";
```
Comment (kyranet) [inline src/classes/console/Console.js]: Then, here:
```javascript
const timestamp = this.timestamps ? `${this.timestamp(`[${moment().format(this.timestamps)}]`, time)} ` : '';
```

File: src/classes/console/Console.js [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    if (this[`_${type}`]) {
+      this[`_${type}`](stuff.split("\n").map(str => `${timestamp}${this.messages(str, message)}`).join("\n"));
+    } else {
+      super.log(stuff.split("\n").map(str => `${timestamp}${this.messages(str, message)}`).join("\n"));
+    }
+  }
+
+  /**
+   * Print something to console as a simple log.
+   * @param  {*} stuff The stuff to log
```
Comment (kyranet) [inline src/classes/console/Console.js]: There's a double space there.

**Review by kyranet — DISMISSED** [review]
A few things before merging this. (I also prefer that each colour could have a type property, and if it doesn't, default to 'log').

File: src/classes/console/Console.js (hunk 2 above)
Comment (CyberiumShadow) [reply src/classes/console/Console.js]: why the value & type equality?

Unless it's to strictly enforce Boolean?

**Review by kyranet — DISMISSED** [review]

**Comment by kyranet** [comment]
Fixes #357

**Pull request merged**